### PR TITLE
Enable search on whole website

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Changes proposed in this pull request:
- Enable Algolia search on whole website

**Related issue(s)**
See also https://github.com/kyma-project/website/issues/307
See also https://github.com/kyma-project/website/issues/350

### CODE REVIEW

**Comment by magicmatatjahu** [comment]
@derberg 
- 7 is limit chosen by me. We only can modified that before initialising 3rd party script from docSearch, so modifications on the fly cannot be done.
- I changed to 2x bigger input width,
- Done
- I did not know that there are such options 😄 We can do some modal window with all options described, for my this is not a problem.
- I thought about that, but I think if we want to make this, we should create follow ticket for that, because this is probably easy to implement, but my task itself should not be related to it.

**Comment by derberg** [comment]
@magicmatatjahu 
* make it 10
* yes, make it bigger
* thank you
* modal will be too much, maybe some kind of tool tip?
* yes, create a followup, will be a killer feature

**Comment by michal-hudy** [comment]
@magicmatatjahu nice work! 👍 

I have noticed that it doesn't look good on mobile devices:
IMG_1253
IMG_E3B994160297-1

File: src/components/layout/Search/Input.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import { InputWrapper, TooltipContent } from "./styled";
+
+interface Props {
+  search?: any;
+}
+
+const Input: React.FunctionComponent<Props> = ({ search }) => {
+  const iconEl = useRef(null);
+  const inputEl = useRef(null);
+  const [inputActive, searchFocus] = useToggle(inputEl, iconEl, false);
```
[the remark this answers is not in the document]
Comment (aerfio) [reply src/components/layout/Search/Input.tsx]: This will solve previous issue with `any` I think

File: src/components/layout/Search/Input.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import { InputWrapper, TooltipContent } from "./styled";
+
+interface Props {
+  search?: any;
+}
+
+const Input: React.FunctionComponent<Props> = ({ search }) => {
+  const iconEl = useRef(null);
+  const inputEl = useRef(null);
```
Comment (aerfio) [inline src/components/layout/Search/Input.tsx]: const inputEl = useRef<HTMLInputElement>(null);

and delete those `as any` assertions

do same thing with iconEl

File: src/components/layout/Search/Input.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  const [inputActive, searchFocus] = useToggle(inputEl, iconEl, false);
+
+  useEffect(() => {
+    focusSearch();
```
Comment (aerfio) [inline src/components/layout/Search/Input.tsx]: `focusSearch` here and `searchFocus` in line 18? change those names please

File: src/components/layout/Search/Input.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        </li>
+      </ul>
+    </TooltipContent>
+  );
+
+  return (
+    <InputWrapper
+      iconName="search"
+      iconPrefix="fas"
+      reference={iconEl}
```
Comment (aerfio) [inline src/components/layout/Search/Input.tsx]: `reference` here and `ref` in line 94
Besides, official docs do not mention `reference` at all

https://reactjs.org/docs/hooks-reference.html#useref

File: src/components/layout/Search/Search.tsx [hunk 5]
```diff
@@ -48,16 +57,15 @@ const Algolia: React.FunctionComponent = () => {
...
     },
     handleSelected,
   };
 
-  const init = (): void => {
+  const loadOptions = (): void => {
     if (initial || !window || !(window as any).docsearch) return;
 
-    (window as any).docsearch(algoliaOptions);
+    search = (window as any).docsearch(algoliaOptions);
```
Comment (aerfio) [inline src/components/layout/Search/Search.tsx]: https://stackoverflow.com/questions/12709074/how-do-you-explicitly-set-a-new-property-on-window-in-typescript

Make new file `window.d.ts` and put thing like this there please, quick search and I found 4 `(window as any).something`

File: src/components/layout/Search/styled.ts [hunk 6]
```diff
@@ -1,82 +1,59 @@
...
+      &:first-child {
+        display: none;
+        color: transparent;
       }
+    `}
 
+    > form {      
       .algolia-autocomplete {
-        margin: 0;
+        margin: 0 !important;
```
Comment (aerfio) [inline src/components/layout/Search/styled.ts]: Is it really neccesary? Can it be done *styled-components* way with `&&&`?

File: src/components/layout/Search/Input.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+          (searchFocus as any)();
+        }
+      }}
+      active={inputActive as boolean}
+    >
+      <form onSubmit={e => e.preventDefault()}>
+        <input
+          id="algolia-search"
+          type="search"
+          placeholder="Search"
```
Comment (aerfio) [inline src/components/layout/Search/Input.tsx]: Don't we have some kind of file with all texts?

Comment (aerfio) [reply src/components/layout/Search/Input.tsx]: If we're planning to have other languages than english in future then we can't do it like this

**Comment by derberg** [comment]
sometimes when you try search it doesn't work and then magically it works again
like for example here https://kyma-project.io/search/ yesterday and today it didn't work and now at the end of the day it works again. What is the reason?

**Comment by aerfio** [comment]
Same thing as @derberg said in my experience

**Comment by aerfio** [comment]
Also: 
image

after searching anything

File: src/common/hooks/useToggle.ts [hunk 8]
```diff
@@ -1,6 +1,10 @@
 import React, { useState, useEffect, useCallback } from "react";
 
-export default function useDropdown(dropEl: any, actionEl: any) {
+export default function useToggle(
+  dropEl: any,
+  actionEl: any,
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/common/hooks/useToggle.ts]: I use React.RefObject. Thanks for that, but you're starting to piss me off with these types XD

File: src/common/hooks/useToggle.ts [hunk 9]
```diff
@@ -1,6 +1,10 @@
...
+  hiddenOnClickOnAction: boolean = true,
+) {
   dropEl = dropEl.current;
   actionEl = actionEl.current;
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/common/hooks/useToggle.ts]: Done

File: src/common/i18n/locales.ts [hunk 10]
```diff
@@ -8,6 +8,7 @@ import enLandingPage from "../../../content/i18n/en/landing-page.json";
 import enDocs from "../../../content/i18n/en/docs.json";
 import enBlog from "../../../content/i18n/en/blog.json";
 import enRoadmap from "../../../content/i18n/en/roadmap.json";
+import enSearch from "../../../content/i18n/en/search.json";
 import enSiteMetadata from "../../../content/i18n/en/siteMetadata.json";
 import enUtils from "../../../content/i18n/en/utils.json";
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/common/i18n/locales.ts]: Done

File: src/common/i18n/types.ts [hunk 11]
```diff
@@ -23,6 +23,7 @@ export interface Internationalization {
     docs: any;
     blog: any;
     roadmap: any;
+    search: any;
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/common/i18n/types.ts]: We can add types for those jsons, but we use library for i18n, which support only string as parameter (as flat map of appropriate object) to render text, so it is the sense to write them?

File: src/components/layout/Layout.tsx [hunk 12]
```diff
@@ -40,8 +43,7 @@ const Layout: React.FunctionComponent<LayoutProps> = ({
   horizontalHeaderBg = true,
   children,
   backToTopButton = true,
-  docSearchLanguage,
-  docSearchVersion,
+  docsMetadata,
   search = false,
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/components/layout/Layout.tsx]: Done

File: src/components/layout/Layout.tsx [hunk 13]
```diff
@@ -53,27 +55,27 @@ const Layout: React.FunctionComponent<LayoutProps> = ({
...
+                  pageTitle={pageTitle}
+                  pageDescription={pageDescription}
+                  siteMetadata={data.site.siteMetadata}
+                />
+                <LayoutWrapper>
+                  <ModalProvider>
+                    <SlidesBanner {...slidesProps} />
+                    <Popup />
+                    <CookiesBanner />
+                    {/* {backToTopButton && <BackToTop />} */}
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/components/layout/Layout.tsx]: Done

File: src/components/layout/Search/Input.tsx (hunk 7 above)
Comment (magicmatatjahu) [reply src/components/layout/Search/Input.tsx]: Done

**Comment by aerfio** [comment]
Just few more things: 
image

I don't know what official designs for this element look like, but maybe add white outline?

And how common would it be that a iPhone 5s user would enter our website? I think current design bring too many problems like these, ask @derberg whether this can be merged and we just say "screw you" to small screen phones
image

**Comment by michal-hudy** [comment]
@magicmatatjahu There are still some bugs on a mobile device:
1. Search for Kyma and then click `x` button - `undefined` is displayed. Also, the popup is still visible:
image
2. Click some search results, Kyma logo is changed:
image
3. Search for anything and click menu or something outside search popup. The popup stays, but the search field is hidden:
image

**Pull request merged**